File: week-06/yc_2468_safe_areas.cs
using System;
using System.Collections.Generic;

class Program {
    static int N;
    static int[,] map;
    static bool[,] visited;
    static int[] dx = { 1, -1, 0, 0 };
    static int[] dy = { 0, 0, 1, -1 };

    static void Main() {
        N = int.Parse(Console.ReadLine());
        map = new int[N, N];
        int maxHeight = 0;

        for (int i = 0; i < N; i++) {
            var line = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
            for (int j = 0; j < N; j++) {
                map[i, j] = line[j];
                if (map[i, j] > maxHeight)
                    maxHeight = map[i, j];
            }
        }

        int maxSafe = 0;
        for (int h = 0; h <= maxHeight; h++) {
            visited = new bool[N, N];
            int count = 0;

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (!visited[i, j] && map[i, j] > h) {
                        BFS(i, j, h);
                        count++;
                    }
                }
            }

            if (count > maxSafe)
                maxSafe = count;
        }

        Console.WriteLine(maxSafe);
    }

    static void BFS(int startX, int startY, int h) {
        Queue<(int, int)> q = new Queue<(int, int)>();
        q.Enqueue((startX, startY));
        visited[startX, startY] = true;

        while (q.Count > 0) {
            var (x, y) = q.Dequeue();

            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];

                if (nx >= 0 && ny >= 0 && nx < N && ny < N) {
                    if (!visited[nx, ny] && map[nx, ny] > h) {
                        visited[nx, ny] = true;
                        q.Enqueue((nx, ny));
                    }
                }
            }
        }
    }
}

File: week-06/yc_2468_safe_areas.java
import java.util.*;

public class Main {
    static int N;
    static int[][] map;
    static boolean[][] visited;
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        map = new int[N][N];

        int maxHeight = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                map[i][j] = sc.nextInt();
                maxHeight = Math.max(maxHeight, map[i][j]);
            }
        }

        int maxSafe = 0;
        for (int h = 0; h <= maxHeight; h++) {
            visited = new boolean[N][N];
            int count = 0;

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (!visited[i][j] && map[i][j] > h) {
                        bfs(i, j, h);
                        count++;
                    }
                }
            }
            maxSafe = Math.max(maxSafe, count);
        }

        System.out.println(maxSafe);
    }

    static void bfs(int startX, int startY, int h) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{startX, startY});
        visited[startX][startY] = true;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0], y = cur[1];

            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];

                if (nx >= 0 && ny >= 0 && nx < N && ny < N) {
                    if (!visited[nx][ny] && map[nx][ny] > h) {
                        visited[nx][ny] = true;
                        q.offer(new int[]{nx, ny});
                    }
                }
            }
        }
    }
}
